crypto/dh/dh_lib.c

```diff
@@ -246,14 +246,21 @@ void DH_get0_pqg(const DH *dh, BIGNUM **p, BIGNUM **q, BIGNUM **g)
 int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 {
     /* q is optional */
-    if (p == NULL || g == NULL)
+    if ((p == NULL && dh->p == NULL)
+        || (g == NULL && dh->g == NULL))
```

Comment:
Just to throw in another option, how about:
````
if (p == dh->p || g == dh->g)
````
That prevents the NULL case (p and dh->p can't both be NULL) and also stops the "resuse the value I got from get0" case.

Referring to the discussion on -dev, I prefer to keep it as illegal to reuse the value obtained from get0 in a subsequent set0 (to do otherwise presents quite a confusing API where transfer of ownership doesn't always occur), but at the same time detects and reports as an error a common case of misuse.